### ISSUE TITLE
BLE HCI is not connecting on CP 7.0.0

### ISSUE THREAD

**Issue by matemaciek** [adafruit/circuitpython#5462]
### CircuitPython version

```python
7.0.0 on Pi Pico with ESP32, connected as in http://helloraspberrypi.blogspot.com/2021/05/raspberry-pi-picocircuitpython_16.html
```

### Code/REPL

```python
As in http://helloraspberrypi.blogspot.com/2021/05/raspberry-pi-picocircuitpython_16.html
```

### Behavior

Same error as in https://github.com/adafruit/nina-fw/issues/26 , despite newest nina-fw (1.7.4). After switching to CP 6.3.0 everything works as expected.

**Comment by AP4130** [comment]
I can confirm this issue; same chips and firmware but different boards.

Adafruit Feather RP2040 with Airlift ESP32 Featherwing, Nina 1.7.4, CP 7.0.0, attempting to start Bluetooth using example code at https://learn.adafruit.com/adafruit-airlift-featherwing-esp32-wifi-co-processor-featherwing/circuitpython-ble with the proper Feather pin settings uncommented.

Resulted in error at line 185 of ESP32.py:
_bleio.BluetoothError: Timeout waiting for HCI response

Unlike OP, reverting to CP 6.3.0 did not fix for me- received the exact same error. Setup confirmed to connect to wifi.

**Comment by PierreDeQuebec** [comment]
Same problem with the Metro M4 AirLift Lite microcontroller. Bluetooth Low Energy works fine on version 6.3.0 but on version 7.0.0, the following error occurs when executing the line "adapter = esp32.start_bluetooth()".

```
Traceback (most recent call last):
  File "code.py", line 53, in <module>
  File "adafruit_airlift/esp32.py", line 193, in start_bluetooth
_bleio.BluetoothError: Timeout waiting for HCI response
```
Line 193 points in the stop_bluetooth method (line 'return').
```

    def stop_bluetooth(self):
        """Stop Bluetooth on the ESP32. Deinitialize the ~busio.UART used for communication"""
        if self._mode != ESP32.BLUETOOTH:
            return
        self._bleio_adapter.enabled = False
        self.reset(ESP32.NOT_IN_USE)
        self._uart.deinit()
        self._uart = None
```

I tried to circumscribe the 'offending' line of code with 'print' commands. But none of the 'print' commands are executed. Obviously, the problem is not in the Python code.

**Comment by LAWBear** [comment]
Same problem with Feather M4 Express with Air Lift ESP32 Coprocessor- breakout board.  Using CP 7.0.0, and have tried both of the adafruit_airlift (mpy & py) library versions.

``code.py output:
Traceback (most recent call last):
  File "code.py", line 50, in <module>
  File "adafruit_airlift/esp32.py", line 193, in start_bluetooth
_bleio.BluetoothError: Timeout waiting for HCI response``

**Comment by nelsonov** [comment]
Happening for me also on Metro M4 Airlift lite using 7.1.0-beta.1.  Exact same error as previous comment.  It's using the code from: https://learn.adafruit.com/adafruit-metro-m4-express-airlift-wifi/circuitpython-ble  I did a direct copy/paste into MU.

```
Auto-reload is on. Simply save files over USB to run them or enter REPL to disable.
code.py output:
Traceback (most recent call last):
  File "code.py", line 50, in <module>
  File "adafruit_airlift/esp32.py", line 193, in start_bluetooth
_bleio.BluetoothError: Timeout waiting for HCI response

Code done running.
```

I've tried with 7.1.0-beta.1 (downloaded) and 7.1.0-beta.1-19-gb83e09858 (compiled).

I'm happy to compile and test if that's of any help to anyone.

**Comment by dhalbert** [comment]
Fixed by #5683.

**Comment by LAWBear** [comment]
**Feather M4 Express with Air Lift ESP32 Coprocessor -breakout:**

I'm sorry to need a bit of help in implementing the kind fix by Dan. I tried the CP 7.1.0-beta with same challenges:

Traceback (most recent call last):
  File "code.py", line 49, in <module>
  File "/lib/adafruit_airlift/esp32.py", line 193, in start_bluetooth
_bleio.BluetoothError: Timeout waiting for HCI response

**Comment by tannewt** [comment]
What beta version? The fix will be in beta.2 that is in the process of being released.

**Comment by LAWBear** [comment]
dafruit CircuitPython 7.1.0-beta.0 on 2021-11-12; Adafruit Feather M4 Express with samd51j19

**Comment by dhalbert** [comment]
This should now be fixed with 7.1.0-beta.2, which is available on https://circuitpython.org.

**Comment by LAWBear** [comment]
kind wizards, I am probably doing something stupid but I seem to have the same response:  Might I be missing an updated library??

Auto-reload is on. Simply save files over USB to run them or enter REPL to disable.
code.py output:
Traceback (most recent call last):
  File "code.py", line 49, in <module>
  File "/lib/adafruit_airlift/esp32.py", line 193, in start_bluetooth
_bleio.BluetoothError: Timeout waiting for HCI response

Code done running.

Press any key to enter the REPL. Use CTRL-D to reload.

Adafruit CircuitPython 7.1.0-beta.2 on 2021-12-10; Adafruit Feather M4 Express with samd51j19

**Comment by dhalbert** [comment]
@LAWBear Assuming you are using an AirLift FeatherWing, what firmware version are you seeing when running this? https://learn.adafruit.com/adafruit-airlift-featherwing-esp32-wifi-co-processor-featherwing/circuitpython-wifi#circuitpython-usage-3085675-6

I tested 7.1.0-beta.2 with  Metro M4 AirLift Lite, running firmware 1.7.4 on the Airlift, and it is working for me.

**Comment by LAWBear** [comment]
Please see following:
code.py output:
ESP32 SPI hardware test
ESP32 found and in idle mode
Firmware vers. bytearray(b'1.7.4\x00')
MAC addr: ['0x70', '0x13', '0x4a', '0x95', '0xab', '0x34']
	2WIRE505_AP		RSSI: -42
	2gIRE505_AP2		RSSI: -45
	ATT7Ypn73P		RSSI: -75
	ATT2M6q963		RSSI: -85
	2WIRE505_AP2		RSSI: -93
Done!

**Comment by dhalbert** [comment]
Could you list which pins you are using, and post your BLE test program?

**Comment by LAWBear** [comment]
From [https://learn.adafruit.com/adafruit-metro-m4-express-airlift-wifi/circuitpython-ble](url)

# If you are using an AirLift Breakout, check that these
# choices match the wiring to your microcontroller board,
# or change them as appropriate.
esp32 = ESP32(
     reset=board.D12,
     gpio0=board.D10,
     busy=board.D11,
     chip_select=board.D13,
     tx=board.TX,
     rx=board.RX,

`import board

from adafruit_ble import BLERadio
from adafruit_ble.advertising.standard import ProvideServicesAdvertisement
from adafruit_ble.services.nordic import UARTService

from adafruit_airlift.esp32 import ESP32

# If you are using a Metro M4 Airlift Lite, PyPortal,
# or MatrixPortal, you can use the default pin settings.
# Leave this DEFAULT line uncommented.
# esp32 = ESP32() # DEFAULT

# If you are using CircuitPython 6.0.0 or earlier,
# on PyPortal and PyPortal Titano only, use the pin settings
# below. Comment out the DEFAULT line above and uncomment
# the line below. For CircuitPython 6.1.0, the pin names
# have changed for these boards, and the DEFAULT line
# above is correct.
# esp32 = ESP32(tx=board.TX, rx=board.RX)

# If you are using an AirLift FeatherWing or AirLift Bitsy Add-On,
# use the pin settings below. Comment out the DEFAULT line above
# and uncomment the lines below.
# If you are using an AirLift Breakout, check that these
# choices match the wiring to your microcontroller board,
# or change them as appropriate.
esp32 = ESP32(
     reset=board.D12,
     gpio0=board.D10,
     busy=board.D11,
     chip_select=board.D13,
     tx=board.TX,
     rx=board.RX,
)

# If you are using an AirLift Shield,
# use the pin settings below. Comment out the DEFAULT line above
# and uncomment the lines below.
# esp32 = ESP32(
#     reset=board.D5,
#     gpio0=board.D6,
#     busy=board.D7,
#     chip_select=board.D10,
#     tx=board.TX,
#     rx=board.RX,
# )

adapter = esp32.start_bluetooth()

ble = BLERadio(adapter)
uart = UARTService()
advertisement = ProvideServicesAdvertisement(uart)

while True:
    ble.start_advertising(advertisement)
    print("waiting to connect")
    while not ble.connected:
        pass
    print("connected: trying to read input")
    while ble.connected:
        # Returns b'' if nothing was read.
        one_byte = uart.read(1)
        if one_byte:
            print(one_byte)
            uart.write(one_byte)`

**Comment by dhalbert** [comment]
@LAWBear OK, I duplicated your exact setup, with a Feather M4 and an AirLift breakout, and I get `waiting to connect`. I originally had the TX and RX wires reversed, and I got `ESP32 did not respond with a startup message`. I did not get `Timeout waiting for HCI response`.

**Comment by LAWBear** [comment]
Mr. Dan, I will reverse my TX/RX & RX/TX wiring to try to duplicate response and advise.

Sincerely appreciate your good efforts.

I have already cleaned-up and loaded new bundles 20211210 and found identical response to various tests, albeit your "version" file worked fine.

**Comment by LAWBear** [comment]
I have "hardened" the Feather M4 Express, Air Lift, GPS, and SD breakout to a non-breadboard assembly to reduce transient SPI and RX/TX interferences (if any).  Additionally, I have tested any BLE modules from the latest "package(s)" with the same HCI response issues.  These issues include "no adapter found" in some of the tested modules.  I have updated the Adafruit ESP32 Coprocessor with the latest known code as 1.7.4 as documentation prescribes.

All other WiFi tests succeed including combined GPS>SD storage tests giving me some assurance that the hardware connections are robust and secure.

![20211213_074313](https://user-images.githubusercontent.com/65572901/145824018-3591049a-e18d-4fb1-9712-e8132550577f.jpg)
![20211213_074347](https://user-images.githubusercontent.com/65572901/145824034-80069311-1e12-4aaf-b1dd-897d56edc04d.jpg)

I remain hopeful for additional diagnostic samples or suggestions for implementation of the BLE capabilities advertised for the Air Lift ESP32 Coprocessor breakout for the Feather M4 Express (et.al.).

Your kind efforts are appreciated.

**Comment by dhalbert** [comment]
@LAWbear can you get it to work with 6.3.0? Your troubles make me think there may be an issue with the Feather or the breakout, since I duplicated your earlier setup pretty much exactly and it worked with my boards.

**Comment by LAWBear** [comment]
Erasing for tidy install, and collection and install of "adafruit-circuitpython-feather_m4_express-en_US-6.3.0.uf2" I seem to have the very same challenge with HCI response. (??)

Traceback (most recent call last):
  File "code.py", line 49, in <module>
  File "adafruit_airlift/esp32.py", line 193, in start_bluetooth
_bleio.BluetoothError: Timeout waiting for HCI response

Re-applying your "version test" it affirms following:

code.py output:
ESP32 SPI hardware test
ESP32 found and in idle mode
Firmware vers. bytearray(b'1.7.4\x00')
MAC addr: ['0x70', '0x13', '0x4a', '0x95', '0xab', '0x34']
	2WIRE505_AP2		RSSI: -41
	2WIRE505_AP		RSSI: -46
	ATT7Ypn73P		RSSI: -69
	This One 5GHz		RSSI: -77
	2WIRE505_AP2		RSSI: -92
	ATT79VgXEa		RSSI: -92
	ATT79VgXEa_Guest		RSSI: -94
Done!

I have re-connected the Air Lift GP0 to Feather M4- D9 (recommended connection for GP0 in BLE instructions) (D10 used for SD module)

I have retained original serial connection from M4>Air Lift as M4/TX>AirLift/RX & M4/RX>AirLift/TX.  You had mentioned your initial test with TX>TX, RX>RX previously. ??   

from Dan;
 **_"I originally had the TX and RX wires reversed, and I got ESP32 did not respond with a startup message. I did not get Timeout waiting for HCI response"_**

**Comment by LAWBear** [comment]
Erasing for tidy install, and collection and install of "adafruit-circuitpython-feather_m4_express-en_US-6.3.0.uf2" and the 6.0 library collections, I seem to have the very same challenge with HCI response. (??)

Traceback (most recent call last):
File "code.py", line 49, in
File "adafruit_airlift/esp32.py", line 193, in start_bluetooth
_bleio.BluetoothError: Timeout waiting for HCI response

Re-applying your "version test" it affirms following:

code.py output:
ESP32 SPI hardware test
ESP32 found and in idle mode
Firmware vers. bytearray(b'1.7.4\x00')
MAC addr: ['0x70', '0x13', '0x4a', '0x95', '0xab', '0x34']
2WIRE505_AP2 RSSI: -41
2WIRE505_AP RSSI: -46
ATT7Ypn73P RSSI: -69
This One 5GHz RSSI: -77
2WIRE505_AP2 RSSI: -92
ATT79VgXEa RSSI: -92
ATT79VgXEa_Guest RSSI: -94
Done!

Adafruit CircuitPython 6.3.0 on 2021-06-01; Adafruit Feather M4 Express with samd51j19

I have re-connected the Air Lift GP0 to Feather M4- D9 (recommended connection for GP0 in BLE instructions) (D10 used for SD module)

I have retained original serial connection from M4>Air Lift as M4/TX>AirLift/RX & M4/RX>AirLift/TX. You had mentioned your initial test with TX>TX, RX>RX previously. ??

from Dan;
"I originally had the TX and RX wires reversed, and I got ESP32 did not respond with a startup message. I did not get Timeout waiting for HCI response"

**Comment by LAWBear** [comment]
It seems I now have some LED blinking behavior not observed previously.  Not exactly sure how to decode, but looks like:

green-blue-3Xyellow-4Xblue  ??????

**Comment by dhalbert** [comment]
Are you seeing errors in the REPL (serial) with 6.3.0)?

The blinking lights codes changed between 6.x and 7.x: https://learn.adafruit.com/welcome-to-circuitpython/troubleshooting#circuitpython-rgb-status-light-2978455-19

Based on the issues you are having, I am thinking there may be a hardware problem, possibly with the AirLift breakout board. You don't have other boards, is that right?

Let us move this discussion to the forums, because at this point I don't think this is a bug, since I cannot reproduce your problems with the same setup. Could you open a thread in https://forums.adafruit.com/viewforum.php?f=60 ? In the thread, could you summarize anything that is working?

It might be time to think about trying other hardware.

**Comment by LAWBear** [comment]
Blinking codes exhibit on non-working modules, so probably historically OK.

I only have one Air Lift module but have recently upgraded to an M4 Express for both speed and compatibility with BLE from Air Lift ESP32 Coprocessor-breakout.

In the previous version 7.1.0, followed by 7.1.0-beta2, the relevant WiFi modules worked just fine on this Adafruit ESP32 Coprocessor-breakout.  Clearly with some assumptions the "hardware" and connections function as advertised with selected modules.  Nothing in the bluetooth/BLE modules has ever responded similarly, with not only the "HCI timeouts", but also "no adapter available".  It's not yet clear to me that another $10 + $7 + 1 week shipping will solve the problem.  

With diligent attention to the instructions, I haven't found any detail on "switching" between WiFi/BLE modes unless that is otherwise embedded in code. (example- do I need to disconnect any SPI controls when BLE mode expecting to use RX/TX and GP0, etc.)  Do I need to change RX>RX & TX>TX or any other ???  Some "echo" testing ??? etc. etc.

I sincerely appreciate your investigations and will prepare comments for forum users who may have found this/similar challenge.

I will revert to latest 7.1.0-beta X with relevant "bundles and continue pursuits.

**Comment by dhalbert** [comment]
@LAWBear When the module is used in WiFi mode, the communication is over the SPI pins. When it is used in BLE mode, the communication is over the TX and RX pins. The module switches modes on reset, based on the state of the CS pin. You can see the code here (and see the whole file, for how the reset is done as well):
https://github.com/adafruit/Adafruit_CircuitPython_AirLift/blob/91740c71c0a0248ca620c78577dda52c7d26a9bf/adafruit_airlift/esp32.py#L118-L121.

The Airlift is product https://www.adafruit.com/product/4201. Are you using that product, or are you using https://www.adafruit.com/product/4172 ?

Here is a photo of my setup, with an AirLift breakout and a Feather M4. It's hard to follow the wiring, but the SPI and TX/RX pins are all hooked up at once.
![IMG_2260](https://user-images.githubusercontent.com/2847802/145905922-447af456-0448-48e1-97cd-4d8b63384085.JPG)

Here is the BLE test program:
```py
import board

from adafruit_ble import BLERadio
from adafruit_ble.advertising.standard import ProvideServicesAdvertisement
from adafruit_ble.services.nordic import UARTService

from adafruit_airlift.esp32 import ESP32

# If you are using a Metro M4 Airlift Lite, PyPortal,
# or MatrixPortal, you can use the default pin settings.
# Leave this DEFAULT line uncommented.
# esp32 = ESP32() # DEFAULT

# If you are using CircuitPython 6.0.0 or earlier,
# on PyPortal and PyPortal Titano only, use the pin settings
# below. Comment out the DEFAULT line above and uncomment
# the line below. For CircuitPython 6.1.0, the pin names
# have changed for these boards, and the DEFAULT line
# above is correct.
# esp32 = ESP32(tx=board.TX, rx=board.RX)

# If you are using an AirLift FeatherWing or AirLift Bitsy Add-On,
# use the pin settings below. Comment out the DEFAULT line above
# and uncomment the lines below.
# If you are using an AirLift Breakout, check that these
# choices match the wiring to your microcontroller board,
# or change them as appropriate.
esp32 = ESP32(
     reset=board.D12,
     gpio0=board.D10,
     busy=board.D11,
     chip_select=board.D13,
     tx=board.TX,
     rx=board.RX,
)

# If you are using an AirLift Shield,
# use the pin settings below. Comment out the DEFAULT line above
# and uncomment the lines below.
# esp32 = ESP32(
#     reset=board.D5,
#     gpio0=board.D6,
#     busy=board.D7,
#     chip_select=board.D10,
#     tx=board.TX,
#     rx=board.RX,
# )

adapter = esp32.start_bluetooth()

ble = BLERadio(adapter)
uart = UARTService()
advertisement = ProvideServicesAdvertisement(uart)

while True:
    ble.start_advertising(advertisement)
    print("waiting to connect")
    while not ble.connected:
        pass
    print("connected: trying to read input")
    while ble.connected:
        # Returns b'' if nothing was read.
        one_byte = uart.read(1)
        if one_byte:
            print(one_byte)
            uart.write(one_byte)
```
This produces `waiting to connect`

Here is the wifi program, which produces successful output.
```py
# SPDX-FileCopyrightText: 2019 ladyada for Adafruit Industries
# SPDX-License-Identifier: MIT

import board
import busio
from digitalio import DigitalInOut
import adafruit_requests as requests
import adafruit_esp32spi.adafruit_esp32spi_socket as socket
from adafruit_esp32spi import adafruit_esp32spi

# Get wifi details and more from a secrets.py file
try:
    from secrets import secrets
except ImportError:
    print("WiFi secrets are kept in secrets.py, please add them there!")
    raise

print("ESP32 SPI webclient test")

TEXT_URL = "http://wifitest.adafruit.com/testwifi/index.html"
JSON_URL = "http://api.coindesk.com/v1/bpi/currentprice/USD.json"


# If you are using a board with pre-defined ESP32 Pins:
#esp32_cs = DigitalInOut(board.ESP_CS)
#esp32_ready = DigitalInOut(board.ESP_BUSY)
#esp32_reset = DigitalInOut(board.ESP_RESET)

# If you have an AirLift Shield:
# esp32_cs = DigitalInOut(board.D10)
# esp32_ready = DigitalInOut(board.D7)
# esp32_reset = DigitalInOut(board.D5)

# If you have an AirLift Featherwing or ItsyBitsy Airlift:
# esp32_cs = DigitalInOut(board.D13)
# esp32_ready = DigitalInOut(board.D11)
# esp32_reset = DigitalInOut(board.D12)

# If you have an externally connected ESP32:
# NOTE: You may need to change the pins to reflect your wiring
esp32_cs = DigitalInOut(board.D13)
esp32_ready = DigitalInOut(board.D11)
esp32_reset = DigitalInOut(board.D12)

spi = busio.SPI(board.SCK, board.MOSI, board.MISO)
esp = adafruit_esp32spi.ESP_SPIcontrol(spi, esp32_cs, esp32_ready, esp32_reset)

requests.set_socket(socket, esp)

if esp.status == adafruit_esp32spi.WL_IDLE_STATUS:
    print("ESP32 found and in idle mode")
print("Firmware vers.", esp.firmware_version)
print("MAC addr:", [hex(i) for i in esp.MAC_address])

for ap in esp.scan_networks():
    print("\t%s\t\tRSSI: %d" % (str(ap["ssid"], "utf-8"), ap["rssi"]))

print("Connecting to AP...")
while not esp.is_connected:
    try:
        esp.connect_AP(secrets["ssid"], secrets["password"])
    except RuntimeError as e:
        print("could not connect to AP, retrying: ", e)
        continue
print("Connected to", str(esp.ssid, "utf-8"), "\tRSSI:", esp.rssi)
print("My IP address is", esp.pretty_ip(esp.ip_address))
print(
    "IP lookup adafruit.com: %s" % esp.pretty_ip(esp.get_host_by_name("adafruit.com"))
)
print("Ping google.com: %d ms" % esp.ping("google.com"))

# esp._debug = True
print("Fetching text from", TEXT_URL)
r = requests.get(TEXT_URL)
print("-" * 40)
print(r.text)
print("-" * 40)
r.close()

print()
print("Fetching json from", JSON_URL)
r = requests.get(JSON_URL)
print("-" * 40)
print(r.json())
print("-" * 40)
r.close()

print("Done!")
```

**Comment by LAWBear** [comment]
Mr. Dan:

With MOST SINCERE appreciation for your exceptional support and detail.  **We** found the issue!

**_Code stopped by auto-reload.
soft reboot_

_Auto-reload is on. Simply save files over USB to run them or enter REPL to disable.
code.py output:
waiting to connect_**

The current hardware seems to be fine, however I had forgotten that the GPS Featherwing is also using the RX/TX.  I guess more determined study is needed to discover IF there MAY be a way to use these two serial modules at the same time.

I verified from your photo the RX/TX setup (which I had), replicated your code samples and response where my **Adafruit 4201** again failed HCI response but WiFi worked fine.

From my "hardened" component assembly I simply un-plugged the Ultimate GPS Featherwing and ran the BLE test and 

![Screenshot_20211214-052232_Bluefruit Connect](https://user-images.githubusercontent.com/65572901/145989375-c3a21684-1546-4771-b794-66465004ac79.jpg)

thank you, thank you, thank you !!!

**Comment by dhalbert** [comment]
Glad you tracked down the problem!

> the GPS Featherwing is also using the RX/TX. I guess more determined study is needed to discover IF there MAY be a way to use these two serial modules at the same time.

You can use some other UART for the GPS, or some other UART for the BLE connection, assuming you have some spare pins. To find out which other pins can be used for a second UART, read about using a small script to determine the available pins here: https://learn.adafruit.com/circuitpython-essentials/circuitpython-uart-serial#wheres-my-uart-2985036-20

**Comment by SirHormigo** [comment]
@dhalbert I'm sorry to refloat this issue. I'm getting the same HCI error with CP 7.3.3 and Nina 1.7.4.

```
"adafruit_airlift/esp32.py", line 193, in start_bluetooth
_bleio.BluetoothError: Timeout waiting for HCI response
```

I've tried multiple CP versions without success: 6.3.0, 7.1.0, 7.1.1, 7.2.1, 7.2.2, 7.3.3 and 8.0.0-beta.4.

Only with version 7.1.1 (I believe it was the first one after the solution merged) I got a different error:

```
"adafruit_ble/__init__.py", line 13, in <module>
ImportError: no module named  '__future__'

```
Please forgive me, but this is the only placed I've been able to locate discussion of this error.

**Comment by lonesoac0** [comment]
I am seeing the samething as @dhalbert 

`code.py output:
Traceback (most recent call last):
  File "code.py", line 49, in <module>
  File "adafruit_airlift/esp32.py", line 186, in start_bluetooth
  File "adafruit_airlift/esp32.py", line 145, in reset
RuntimeError: ESP32 did not respond with a startup message`

**Comment by dhalbert** [comment]
Closing this for now. Please comment if you have further problems.
